### ISSUE TITLE
Uninitialized constant Gem::Specification under JRuby

### ISSUE THREAD

**Issue by liveh2o** [bundler/bundler#5248]
- What did you do?
I ran the command `/usr/local/rvm/gems/jruby-9.0.1.0@global/bin/bundle install --gemfile /srv/firefly/releases/20161207190858/Gemfile --path /srv/firefly/shared/bundle --deployment --quiet --without development test`

- What did you expect to happen?
I expected Bundler to install the bundle on the server.

- What happened instead?
Instead, what actually happened was an exception was raised.

It's worth noting that this only happens when deploying to production. It's also intermittent in that different servers experience the issue at different times and deploying again to the same server works just fine.

Error details
```
NameError: uninitialized constant Gem::Specification
org/jruby/RubyModule.java:3166:in `const_missing'
/usr/local/rvm/gems/jruby-9.0.1.0@global/gems/bundler-1.12.5/lib/bundler/remote_specification.rb:78:in `method_missing'
/usr/local/rvm/gems/jruby-9.0.1.0@global/gems/bundler-1.12.5/lib/bundler/installer/parallel_installer.rb:59:in `all_dependencies'
/usr/local/rvm/gems/jruby-9.0.1.0@global/gems/bundler-1.12.5/lib/bundler/installer/parallel_installer.rb:47:in `dependencies'
/usr/local/rvm/gems/jruby-9.0.1.0@global/gems/bundler-1.12.5/lib/bundler/installer/parallel_installer.rb:40:in `dependencies_installed?'
/usr/local/rvm/gems/jruby-9.0.1.0@global/gems/bundler-1.12.5/lib/bundler/installer/parallel_installer.rb:120:in `block in enqueue_specs'
org/jruby/RubyArray.java:1560:in `each'
/usr/local/rvm/gems/jruby-9.0.1.0@global/gems/bundler-1.12.5/lib/bundler/installer/parallel_installer.rb:119:in `enqueue_specs'
/usr/local/rvm/gems/jruby-9.0.1.0@global/gems/bundler-1.12.5/lib/bundler/installer/parallel_installer.rb:81:in `call'
/usr/local/rvm/gems/jruby-9.0.1.0@global/gems/bundler-1.12.5/lib/bundler/installer/parallel_installer.rb:64:in `call'
/usr/local/rvm/gems/jruby-9.0.1.0@global/gems/bundler-1.12.5/lib/bundler/installer.rb:171:in `install_in_parallel'
/usr/local/rvm/gems/jruby-9.0.1.0@global/gems/bundler-1.12.5/lib/bundler/installer.rb:156:in `install'
/usr/local/rvm/gems/jruby-9.0.1.0@global/gems/bundler-1.12.5/lib/bundler/installer.rb:71:in `run'
/usr/local/rvm/gems/jruby-9.0.1.0@global/gems/bundler-1.12.5/lib/bundler/installer.rb:22:in `install'
/usr/local/rvm/gems/jruby-9.0.1.0@global/gems/bundler-1.12.5/lib/bundler/cli/install.rb:102:in `run'
/usr/local/rvm/gems/jruby-9.0.1.0@global/gems/bundler-1.12.5/lib/bundler/cli.rb:175:in `install'
/usr/local/rvm/gems/jruby-9.0.1.0@global/gems/bundler-1.12.5/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/usr/local/rvm/gems/jruby-9.0.1.0@global/gems/bundler-1.12.5/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/usr/local/rvm/gems/jruby-9.0.1.0@global/gems/bundler-1.12.5/lib/bundler/vendor/thor/lib/thor.rb:359:in `dispatch'
/usr/local/rvm/gems/jruby-9.0.1.0@global/gems/bundler-1.12.5/lib/bundler/vendor/thor/lib/thor/base.rb:440:in `start'
/usr/local/rvm/gems/jruby-9.0.1.0@global/gems/bundler-1.12.5/lib/bundler/cli.rb:11:in `start'
/usr/local/rvm/gems/jruby-9.0.1.0@global/gems/bundler-1.12.5/exe/bundle:27:in `block in (root)'
/usr/local/rvm/gems/jruby-9.0.1.0@global/gems/bundler-1.12.5/lib/bundler/friendly_errors.rb:98:in `with_friendly_errors'
/usr/local/rvm/gems/jruby-9.0.1.0@global/gems/bundler-1.12.5/exe/bundle:19:in `<top>'
org/jruby/RubyKernel.java:957:in `load'
/usr/local/rvm/gems/jruby-9.0.1.0@global/bin/bundle:1:in `<eval>'
org/jruby/RubyKernel.java:978:in `eval'
/usr/local/rvm/gems/jruby-9.0.1.0@global/bin/jruby_executable_hooks:15:in `<top>'
```
Environment
```
Bundler   1.12.5
Rubygems  2.4.8
Ruby      2.2.2p0 (2015-09-02 revision 50293) [java]
GEM_HOME  /srv/firefly/shared/bundle/jruby/2.2.0
GEM_PATH
RVM       1.26.9 (1.26.9)
Git       1.9.1
```
Bundler settings
```
frozen
Set for your local app (/srv/firefly/releases/20161207190858/.bundle/config): "1"
path
Set for your local app (/srv/firefly/releases/20161207190858/.bundle/config): "/srv/firefly/shared/bundle"
without
Set for your local app (/srv/firefly/releases/20161207190858/.bundle/config): "development:test"
disable_shared_gems
Set for your local app (/srv/firefly/releases/20161207190858/.bundle/config): true
orig_path
Set via BUNDLE_ORIG_PATH: "/usr/local/rvm/gems/jruby-9.0.1.0@firefly/bin:/usr/local/rvm/gems/jruby-9.0.1.0@global/bin:/usr/local/rvm/rubies/jruby-9.0.1.0/bin:/usr/local/rvm/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games"
orig_gem_path
Set via BUNDLE_ORIG_GEM_PATH: "/usr/local/rvm/gems/jruby-9.0.1.0@firefly:/usr/local/rvm/gems/jruby-9.0.1.0@global"
gemfile
Set via BUNDLE_GEMFILE: "/srv/firefly/releases/20161207190858/Gemfile"
```

**Comment by segiddins** [comment]
I'm confused how `rubygems` could have not been required by that point -- @liveh2o can you try monkey-patching bundler to print out what `$LOADED_FEATURES` is where that exception is raised? Thanks!

**Comment by liveh2o** [comment]
Unfortunately, that's not really doable since we are seeing this with deploying to our production servers. I'm completely confused as to how `rubygems` is not required. Based on the fact that this does not always happen (i.e. it works most of the time), it would appear to be some sort of concurrency issue.

Additionally, we are deploying other applications to multiple servers at the same time without seeing this problem, so it may be related to JRuby 9.0.1.0.

**Comment by segiddins** [comment]
Could you also try updating to the latest release of bundler (1.14.0.pre.2) to see if that happens to fix the issue? Thanks!

**Comment by liveh2o** [comment]
Since 1.14.x has been released and after upgrading to JRuby 9.1.7.0, I'm no longer seeing the issue. I'm going to close this. Thanks for the assistance!
